### PR TITLE
VReplication: throttling info for both source and target; Online DDL propagates said info

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

This PR is primarily an enhancement to VReplication, adding visibility into throttling status of an active workflow.

In addition, Online DDL now utilizes said information and presents it as part of `SHOW VITESS_MIGRATIONS`.

Lastly, there's a deadlock solution to a particular throttling scenario.

### Throttling visibility in VReplication

The following components in VReplication consult the throttler (way before this PR):

- `vstreamer`, on source tablet
- `rowstreamer`, on source tablet
- `vplayer`, on target tablet
- `vcopier`, on target tablet

As of this PR, all four further advertise their throttling state. The table `_vt.vreplication` now has two new columns:

- `time_throttled` (unix time, seconds)
- `component_throttled` (name of component)

These present the last known incident where a VReplication component was throttled. Example values might be:
```
             time_throttled: 1656420364
        component_throttled: vcopier
```

For `vplayer` and `vcopier`, whenever they are throttled, they now ask `vreplicator` to update those values, e.g. via:
```go
vr.updateTimeThrottled("vplayer")
```

`vreplicator` rate-limits those writes, to at most one per second. It's fine if calls to `updateTimeThrottled()` are frequent. Some calls may be dropped. The database will only be affected at most once per second.

As for `vstreamer` and `rowstreamer`, they don't have immediate access to `_vt.vreplication` or to `vreplicator`, because they are on the source tablet. A new `bool throttled` proto field is added in both `VEvent` and `VStreamRowsResponse`. If `vstreamer` is throttled, it sends (rate limited) heartbeat events with `Throttled: true`. If `VStreamRowsResponse` is throttled, it sends (rate limited) responses where `Throttled: true`. `vplayer` and `vcopier` receive and identify those responses, respectively, and call `updateTimeThrottled()`.

This means we can see in `_vt.vreplication` the last known throttling incident: when it happened, and which component was affected, one of `vstreamer, rowstreamer, vplayer, vcopier`.

This information does _not_ include frequency of throttling, number of successful/rejected throttler checks, etc. We leave that to a future PR.

### Online DDL throttling info

When reviewing running migrations, and for `vitess` migration, the online DDL executor now reads the above two columns, and propagate them as two new columns in `_vt.schema_migrations` table:

- `last_throttled_timestamp` (`TIMESTAMP`)
- `component_throttled` (textual)

These are only sampled once per minute, sometimes more frequently.

### A deadlock scenario, now solved

With the introduction of [on demand heartbeats](https://github.com/vitessio/vitess/pull/10198), we've introduced a throttling deadlock scenario. Consider the following scenario (Copy of comment from this PR):

- a vitess migration
- with on demand heartbeats
- the streamer running on a replica
- the streamer was throttled for long enough
- then vplayer and vcopier are blocked, waiting for the streamer to do something
- since they are blocked, they're not running throttler checks
- since streamer runs on replica, it only checks that replica
- therefore no one asking for on-demand heartbeats
- then, if the conditions for the streamer's throttling are done, the streamer then thinks there's replication lag, with nothing to remediate it.
- it's a deadlock.

And so, once per `reviewRunningMigrations()`, and assuming there _are_ running migrations, we ensure to hit a throttler check. This will kick on-demand heartbeats, unlocking the deadlock.

### rowstreamer heartbeats

`rowstreamer` now sends heartbeats every `10sec`. This means even if it's blocked on reading the next batch of rows (blocked, slow, hangs, for whatever reason), it still reports heartbeats. these are intercepted by `vcopier` which then updates `_vt.vreplication.time_heartbeat`.

## Related Issue(s)

#6926 
#10198
 
## Checklist

-   [ ] "Backport me!" label has been added if this change should be backported
-   [x] Tests were added or are not required
-   [ ] Documentation was added or is not required

### CODE REVIEW

**Comment by shlomi-noach** [comment]
I forgot to mention. One of the side effects to this PR, which is actually the root trigger to working on it:

Without this PR, a VReplication workflow that was throttled, would be identified by the Online DDL executor as "stale" after 10 minutes of inactivity, and terminated.

Now, with this PR, this does not happen because `time_updated`  in `_vt.vreplication` is updated whenever a throttling is identified. Thus, Online DDL executor keeps seeing the migration as "fresh", and does not terminate it.

**Comment by shlomi-noach** [comment]
Looking into unit test failures, which seem to be related to the `RateLimiter`.

**Comment by shlomi-noach** [comment]
Obviously only failing in GitHub CI :stuck_out_tongue:  and not on local envs.

**Comment by shlomi-noach** [comment]
Added the tow new columns in `ReplicationStatusResult`, which then shows up in `vtctl Workflow ... Show`

**Comment by shlomi-noach** [comment]
Also added:

`rowstreamer` now sends heartbeats every `10sec`. This means even if it's blocked on reading the next batch of rows (blocked, slow, hangs, for whatever reason), it still reports heartbeats. these are intercepted by `vcopier` which then updates `_vt.vreplication.time_heartbeat`.

**Comment by shlomi-noach** [comment]
Fixed all unit tests, new ones showing! Still working through them.

**Comment by shlomi-noach** [comment]
Phew, the tests were brutal but legit! Now good to go!

File: go/vt/vttablet/onlineddl/vrepl.go
Comment (mattlord) [inline go/vt/vttablet/onlineddl/vrepl.go:69]: `Updating` any of them, right?

File: go/vt/vttablet/tabletmanager/vreplication/vcopier.go
Comment (mattlord) [inline go/vt/vttablet/tabletmanager/vreplication/vcopier.go:242]: I think this could be a resultstreamer too, but this is precise enough (I'm thinking of Materialization and VDiff1), but I could be wrong too). If that's true, and we care to distinguish them here, we could infer this from workflow type etc.

File: go/vt/vttablet/tabletmanager/vreplication/vreplicator.go
Comment (mattlord) [inline go/vt/vttablet/tabletmanager/vreplication/vreplicator.go:508]: I feel we it may be worth having constants defined for the components, e.g. `vreplication.Compenent` type (alias for string) with a value of e.g. `vreplication.VPlayerComponent` being a const string for "vplayer"? Strings introduces the opportunity for typos, invalid components, etc.

File: go/vt/vttablet/tabletserver/vstreamer/vstreamer.go
Comment (mattlord) [inline go/vt/vttablet/tabletserver/vstreamer/vstreamer.go:281]: I'm curious why we get a nanosecond precision timestamp only to turn it back into a second precision timestamp for the protobuf message?

**Review by mattlord — COMMENTED** [review]
I like it! This will greatly improve the observability/debugability of VReplication workflows (especially when we enable tablet throttling by default). Thank you for working on it!

I had a few questions and comments, but nothing major. I'll look out for your responses and can approve quickly if needed.

@rohit-nayak-ps if you have time, would you mind giving it a pass too? Thanks!

File: go/vt/vttablet/tabletmanager/vreplication/vcopier.go
Comment (rohit-nayak-ps) [reply go/vt/vttablet/tabletmanager/vreplication/vcopier.go:242]: > I think this could be a resultstreamer too, but this is precise enough (I'm thinking of Materialization and VDiff1), but I could be wrong too). If that's true, and we care to distinguish them here, we could infer this from workflow type etc.

Only VDiff1 uses resultstreamer and it doesn't use the vcopier, so this is good

**Review by rohit-nayak-ps — APPROVED** [review]
lgtm

File: go/vt/vttablet/tabletserver/vstreamer/vstreamer.go
Comment (shlomi-noach) [reply go/vt/vttablet/tabletserver/vstreamer/vstreamer.go:281]: I'm not sure, this block of code is moved from elsewhere and otherwise untouched.

File: go/vt/binlog/binlogplayer/binlog_player.go
Comment (shlomi-noach) [reply go/vt/binlog/binlogplayer/binlog_player.go:571]: Absolutely right. Too many copy_pastes.

File: go/vt/vttablet/onlineddl/executor.go
Comment (shlomi-noach) [reply go/vt/vttablet/onlineddl/executor.go:3239]: typo for `locaked`. Fixed.

File: go/vt/binlog/binlogplayer/binlog_player.go
Comment (shlomi-noach) [reply go/vt/binlog/binlogplayer/binlog_player.go]: Renamed as `timeThrottledUnix`

File: go/vt/binlog/binlogplayer/binlog_player.go
Comment (shlomi-noach) [reply go/vt/binlog/binlogplayer/binlog_player.go]: fixed

File: go/vt/vttablet/onlineddl/schema.go
Comment (shlomi-noach) [reply go/vt/vttablet/onlineddl/schema.go:81]: We're running out on `varchar` capacity in `_vt.schema_migrations`, which is why I now use `*text` types for everything. But, I'm curious, isn't `tinytext` an equivalent of `varchar(255)`? And it can be indexed just as with a `varchar`? At least I've always indexed `*text` types without a problem.

File: go/vt/vttablet/onlineddl/schema.go
Comment (shlomi-noach) [reply go/vt/vttablet/onlineddl/schema.go:103]: So just to clarify, the column type in this case is a `TIMESTAMP` and there was a long standing bug where we erroneously `INSERT`ed a  `FROM_UNIXTIME()` value, which of course did not make any sense.

We are now looking at a 1.5 year old decision, to use `TIMESTAMP` values in `_vt.schema_migrations` table, long before it was even associated with `_vt.vreplication`, and long before I knew of th eexistence of `_vt.vreplication`. This is how the code "grew" and now it's too late to change `_vt.schema_migrations` from `TIMESTAMP` and into `INT`.

The fact that we copy values from `_vt.vreplication`, that happen to be `INT` types, and into `_vt.schema_migrations` with `TIMESTAMP` types, is unfortunate, but also does not tell the entire story. For long, `schema_migrations` acted with `gh-ost` and `pt-osc`, oblivious of `vreplication`.

File: go/vt/vttablet/onlineddl/vrepl.go
Comment (shlomi-noach) [reply go/vt/vttablet/onlineddl/vrepl.go:69]: I didn't know that, but you're the native English speaker :sweat_smile:

File: go/test/endtoend/onlineddl/vrepl/onlineddl_vrepl_test.go
Comment (shlomi-noach) [reply go/test/endtoend/onlineddl/vrepl/onlineddl_vrepl_test.go]: Refactored and now using a `WaitForThrottledTimestamp()` function.

File: go/test/endtoend/onlineddl/vrepl/onlineddl_vrepl_test.go
Comment (shlomi-noach) [reply go/test/endtoend/onlineddl/vrepl/onlineddl_vrepl_test.go]: Refactored and now using a `WaitForThrottledTimestamp()` function.

File: go/vt/vttablet/tabletmanager/vreplication/vreplicator.go
Comment (shlomi-noach) [reply go/vt/vttablet/tabletmanager/vreplication/vreplicator.go:510]: Fixed. Now formalized as `VReplicationComponent`.

Comment (shlomi-noach) [reply go/vt/vttablet/tabletmanager/vreplication/vreplicator.go]: Rather, `ComponentName`.

File: go/vt/vttablet/tabletserver/vstreamer/vstreamer.go
Comment (mattlord) [reply go/vt/vttablet/tabletserver/vstreamer/vstreamer.go:281]: OK, let's assume there was some good reason. 😄

**Review by mattlord — APPROVED** [review]
Looks great! Thank you ❤️

**Pull request merged**
